File: packages/mobile_android/app/src/main/java/dev/verdeai/app/BrowseScreens.kt
package dev.verdeai.app

import androidx.compose.foundation.ExperimentalFoundationApi
import androidx.compose.foundation.background
import androidx.compose.foundation.combinedClickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.LazyListScope
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material3.*
import androidx.compose.material3.pulltorefresh.PullToRefreshBox
import androidx.compose.runtime.*
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalClipboardManager
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import dev.verdeai.core.*
import kotlinx.coroutines.delay

/** Wall clock for elapsed timers; tests pass a fixed clock with ticking disabled. */
internal class UiClock(val now: () -> Long = System::currentTimeMillis, val ticking: Boolean = true)
internal const val BROWSE_LIST = "browse-list"
internal val LocalUiClock = staticCompositionLocalOf { UiClock() }

@Composable
internal fun rememberNow(enabled: Boolean): Long {
    val clock = LocalUiClock.current
    val now by produceState(clock.now(), enabled, clock) {
        value = clock.now()
        while (enabled && clock.ticking) { delay(1_000); value = clock.now() }
    }
    return now
}

// ---- Pure presentation rules (unit tested) ----

internal fun statusLabel(status: String): String = when (status) {
    "waiting_approval" -> "Needs approval"
    "waiting" -> "Waiting"
    "working", "running", "accepted" -> "Working"
    "idle" -> "Idle"
    "exited" -> "Exited"
    "unavailable" -> "Unavailable"
    "completed" -> "Done"
    "failed" -> "Failed"
    "aborted", "cancelled" -> "Stopped"
    else -> status.replace('_', ' ').replaceFirstChar { it.uppercase() }
}

/** K-17 attention reason shown as a badge; null when the core reports none. */
internal fun attentionLabel(kind: String?): String? = when (kind) {
    null -> null
    "unread" -> "Unread"
    "needs_approval" -> "Needs approval"
    "blocked" -> "Blocked"
    "failed" -> "Failed"
    else -> kind.replace('_', ' ').replaceFirstChar { it.uppercase() }
}

internal fun paneLabel(pane: Pane): String = when {
    pane.kind == "terminal" && pane.status == "working" -> "Running"
    else -> statusLabel(pane.status)
}

internal fun elapsedLabel(startMs: Long, nowMs: Long): String {
    val total = ((nowMs - startMs).coerceAtLeast(0)) / 1000
    val h = total / 3600; val m = (total % 3600) / 60; val s = total % 60
    return if (h > 0) "%d:%02d:%02d".format(h, m, s) else "%d:%02d".format(m, s)
}

internal fun agoLabel(thenMs: Long, nowMs: Long): String {
    val minutes = (nowMs - thenMs).coerceAtLeast(0) / 60_000
    return when {
        minutes < 1 -> "just now"
        minutes < 60 -> "$minutes min ago"
        minutes < 24 * 60 -> "${minutes / 60} h ago"
        else -> "${minutes / (24 * 60)} d ago"
    }
}

internal fun paneLine(pane: Pane, nowMs: Long): String {
    val kind = when (pane.kind) { "terminal" -> "Terminal"; "browser" -> "Browser"; else -> "Chat" }
    val timer = pane.started_at_ms?.takeIf { pane.can_stop }?.let { " · ${elapsedLabel(it, nowMs)}" } ?: ""
    return "$kind · ${paneLabel(pane)}$timer"
}

internal fun subagent(thread: ThreadSummary) = thread.thread_id.startsWith("subagent:")

/** Most recent non-archived top-level chats from the core's history projection. */
internal fun recentThreads(workspaces: WorkspacesView?, limit: Int = 8): List<ThreadSummary> =
    workspaces?.history?.items.orEmpty().filter { !it.archived && !subagent(it) }.take(limit)

internal fun workspaceThreads(workspace: Workspace): List<ThreadSummary> =
    workspace.threads.filter { !subagent(it) }.sortedWith(compareByDescending<ThreadSummary> { it.last_activity_at_ms ?: 0 }.thenBy { it.thread_id })

internal fun openable(pane: Pane) = when (pane.kind) {
    "chat" -> pane.thread_id != null
    "terminal" -> pane.terminal_id != null
    else -> false
}

internal enum class BannerAction { Retry, Hosts }
internal data class Banner(val text: String, val action: BannerAction? = null, val error: Boolean = false, val busy: Boolean = false)

/** Cached views, or a live projection backed by at least one snapshot. */
internal fun hasContent(state: BrowseState) = state.hasData &&
    (state.savedAtMs != null || state.host?.sync_state in setOf("ready", "stale"))

internal fun showSpinner(state: BrowseState): Boolean {
    val host = state.host ?: return !state.fatal
    return state.networkAvailable && !state.fatal && state.row?.fatal != true &&
        host.phase != "failed" && host.error == null && host.trust_proposal == null && !host.update_required
}

internal fun needsPairing(state: BrowseState) = state.host?.auth_state in setOf("unpaired", "signed_out", "signing_out")

internal fun browseBanner(state: BrowseState, nowMs: Long): Banner? {
    val host = state.host
    val saved = state.savedAtMs?.let { " Showing saved data from ${agoLabel(it, nowMs)}." } ?: ""
    return when {
        state.fatal || state.row?.fatal == true -> Banner("Connection unavailable — reopen Verde.", error=true)
        host == null || host.auth_state == "loading" -> if (state.savedAtMs != null) Banner("Loading.$saved", busy=true) else null
        needsPairing(state) -> null
        host.auth_state == "repair_required" -> Banner("Pair again — device authorization needs renewal.", BannerAction.Hosts, error=true)
        host.trust_proposal != null -> Banner("Review this host's identity before connecting.", BannerAction.Hosts, error=true)
        host.update_required -> Banner("Update required — update Verde on this phone or host.", error=true)
        !state.networkAvailable -> Banner("You're offline.$saved", error=true)
        host.phase == "failed" || host.error?.failure_kind == "network" ->
            Banner("Can't reach ${host.label}. Is Tailscale on?$saved", BannerAction.Retry, error=true)
        state.home?.error != null || state.workspaces?.error != null ->
            Banner("Couldn't load the latest data. Pull down to retry.$saved", BannerAction.Retry, error=true)
        host.phase != "ready" -> Banner("Connecting…$saved", busy=true)
        state.savedAtMs != null -> Banner("Updating…$saved", busy=true)
        state.home?.incomplete_scopes?.isNotEmpty() == true -> Banner("Some host data is still loading.")
        else -> null
    }
}

// ---- Shared pieces ----

@Composable
internal fun statusColor(status: String, attention: Boolean): Color {
    val colors = MaterialTheme.colorScheme
    return when {
        attention || status == "waiting_approval" -> colors.error
        status in setOf("working", "running", "accepted", "waiting") -> colors.primary
        else -> colors.outline
    }
}

@Composable
private fun Dot(color: Color, description: String) {
    Box(Modifier.size(10.dp).background(color, CircleShape).semantics { contentDescription = description })
}

@OptIn(ExperimentalFoundationApi::class)
@Composable
private fun MenuRow(
    headline: String,
    supporting: String?,
    enabled: Boolean,
    onOpen: () -> Unit,
    menu: List<Pair<String, () -> Unit>>,
    leading: (@Composable () -> Unit)? = null,
    badge: String? = null,
) {
    var expanded by remember { mutableStateOf(false) }
    Box {
        ListItem(
            headlineContent = { Text(headline, maxLines = 1, overflow = TextOverflow.Ellipsis) },
            supportingContent = supporting?.let { { Text(it, maxLines = 2, overflow = TextOverflow.Ellipsis) } },
            leadingContent = leading,
            trailingContent = badge?.let { { AttentionBadge(it) } },
            modifier = Modifier.fillMaxWidth().combinedClickable(
                enabled = enabled || menu.isNotEmpty(),
                onClickLabel = "Open", onLongClickLabel = "More actions",
                onClick = { if (enabled) onOpen() else if (menu.isNotEmpty()) expanded = true },
                onLongClick = { if (menu.isNotEmpty()) expanded = true }),
        )
        DropdownMenu(expanded = expanded, onDismissRequest = { expanded = false }) {
            menu.forEach { (label, action) ->
                DropdownMenuItem(text = { Text(label) }, onClick = { expanded = false; action() })
            }
        }
    }
}

@Composable
private fun AttentionBadge(text: String) {
    Badge(containerColor = MaterialTheme.colorScheme.errorContainer, contentColor = MaterialTheme.colorScheme.onErrorContainer) {
        Text(text, Modifier.padding(horizontal = 4.dp))
    }
}

@Composable
private fun PaneItem(pane: Pane, now: Long, onOpen: (Pane) -> Unit) {
    @Suppress("DEPRECATION") val clipboard = LocalClipboardManager.current
    val canOpen = openable(pane)
    val menu = buildList {
        if (canOpen) add("Open" to { onOpen(pane) })
        add("Copy title" to { clipboard.setText(AnnotatedString(pane.title)) })
    }
    MenuRow(pane.title, paneLine(pane, now), canOpen, { onOpen(pane) }, menu,
        leading = { Dot(statusColor(pane.status, pane.attention), paneLabel(pane)) },
        badge = attentionLabel(pane.attention_kind))
}

@Composable
private fun ThreadItem(thread: ThreadSummary, now: Long, workspaceLabel: String?, onOpen: (ThreadSummary) -> Unit) {
    @Suppress("DEPRECATION") val clipboard = LocalClipboardManager.current
    val parts = listOfNotNull(workspaceLabel, statusLabel(thread.status).takeIf { thread.status != "idle" },
        thread.last_activity_at_ms?.let { agoLabel(it, now) })
    MenuRow(thread.title, parts.joinToString(" · ").ifEmpty { null }, true, { onOpen(thread) },
        listOf("Open" to { onOpen(thread) }, "Copy title" to { clipboard.setText(AnnotatedString(thread.title)) }),
        leading = { Dot(statusColor(thread.status, thread.status == "waiting_approval"), statusLabel(thread.status)) })
}

private fun LazyListScope.header(text: String) {
    item(key = "header:$text") {
        Text(text, style = MaterialTheme.typography.titleSmall, color = MaterialTheme.colorScheme.primary,
            modifier = Modifier.padding(start = 16.dp, end = 16.dp, top = 20.dp, bottom = 4.dp))
    }
}

private fun LazyListScope.note(key: String, text: String) {
    item(key = key) { Text(text, Modifier.padding(horizontal = 16.dp, vertical = 8.dp), style = MaterialTheme.typography.bodyMedium) }
}

/** Title, host status, banner and pull-to-refresh shared by the browse screens. */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun BrowseFrame(
    title: String,
    state: BrowseState,
    now: Long,
    onRefresh: () -> Unit,
    onHosts: () -> Unit,
    onBack: (() -> Unit)? = null,
    content: LazyListScope.() -> Unit,
) {
    Column(Modifier.fillMaxSize()) {
        if (onBack != null) {
            TopAppBar(title = { Text(title, maxLines = 1, overflow = TextOverflow.Ellipsis) },
                navigationIcon = { IconButton(onClick = onBack) { Icon(Icons.AutoMirrored.Filled.ArrowBack, contentDescription = "Back") } })
        }
        PullToRefreshBox(isRefreshing = state.refreshing, onRefresh = onRefresh, modifier = Modifier.fillMaxSize()) {
            LazyColumn(Modifier.fillMaxSize().testTag(BROWSE_LIST)) {
                if (onBack == null) item(key = "title") {
                    Column(Modifier.padding(start = 16.dp, end = 16.dp, top = 16.dp)) {
                        Text(title, style = MaterialTheme.typography.headlineLarge)
                        state.row?.let { row ->
                            Row(verticalAlignment = androidx.compose.ui.Alignment.CenterVertically,
                                horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                                val status = hostStatus(row)
                                Dot(hostDotColor(row), status)
                                Text("${row.saved.label} · $status", style = MaterialTheme.typography.bodyMedium)
                            }
                        }
                    }
                }
                browseBanner(state, now)?.let { banner ->
                    item(key = "banner") {
                        Card(Modifier.fillMaxWidth().padding(horizontal = 16.dp, vertical = 8.dp),
                            colors = CardDefaults.cardColors(containerColor =
                                if (banner.error) MaterialTheme.colorScheme.errorContainer else MaterialTheme.colorScheme.secondaryContainer)) {
                            Column(Modifier.padding(12.dp), verticalArrangement = Arrangement.spacedBy(8.dp)) {
                                Text(banner.text)
                                if (banner.busy) LinearProgressIndicator(Modifier.fillMaxWidth())
                                when (banner.action) {
                                    BannerAction.Retry -> TextButton(onClick = onRefresh, enabled = !state.refreshing) { Text("Retry") }
                                    BannerAction.Hosts -> TextButton(onClick = onHosts) { Text("Open hosts") }
                                    null -> Unit
                                }
                            }
                        }
                    }
                }
                content()
            }
        }
    }
}

/** Shared host-level gates; returns true when the caller should render its own content. */
private fun LazyListScope.gate(state: BrowseState, onPair: () -> Unit): Boolean {
    when {
        state.hostId == null -> {
            note("no-host", "No host selected.")
            item(key = "choose") { TextButton(onClick = onPair, modifier = Modifier.padding(horizontal = 8.dp)) { Text("Choose a host") } }
        }
        needsPairing(state) -> {
            note("unpaired", "This phone isn't paired with ${state.row?.saved?.label ?: "this host"} yet.")
            item(key = "pair") { Button(onClick = onPair, modifier = Modifier.padding(horizontal = 16.dp)) { Text("Pair with this host") } }
        }
        !hasContent(state) -> {
            if (showSpinner(state)) item(key = "loading") {
                Column(Modifier.fillMaxWidth().padding(32.dp), horizontalAlignment = androidx.compose.ui.Alignment.CenterHorizontally,
                    verticalArrangement = Arrangement.spacedBy(12.dp)) {
                    CircularProgressIndicator()
                    Text("Loading workspaces…")
                }
            } else note("nodata", "No data yet. Pull down to retry when the host is reachable.")
        }
        else -> return true
    }
    return false
}

// ---- Screens ----

@Composable
internal fun HomeScreen(
    model: BrowseModel,
    onOpenPane: (Pane) -> Unit,
    onOpenThread: (ThreadSummary) -> Unit,
    onOpenWorkspace: (String) -> Unit,
    onHosts: () -> Unit,
    onPair: () -> Unit,
) {
    val state by model.state.collectAsState()
    val panes = state.home?.items.orEmpty()
    val now = rememberNow(panes.any { it.can_stop && it.started_at_ms != null })
    BrowseFrame("Home", state, now, model::refresh, onHosts) {
        if (!gate(state, onPair)) return@BrowseFrame
        val attention = panes.filter { it.attention }
        val running = panes.filterNot { it.attention }
        if (attention.isNotEmpty()) {
            header("Needs attention")
            items(attention, key = { "attention:" + it.id }) { PaneItem(it, now, onOpenPane) }
        }
        if (running.isNotEmpty()) {
            header("Running")
            items(running, key = { "running:" + it.id }) { PaneItem(it, now, onOpenPane) }
        }
        if (panes.isEmpty()) note("idle", "Nothing is running or waiting on you.")
        val labels = state.workspaces?.items.orEmpty().associate { it.workspace_id to it.label }
        val recent = recentThreads(state.workspaces)
        if (recent.isNotEmpty()) {
            header("Recent chats")
            items(recent, key = { "recent:${it.workspace_id}:${it.thread_id}" }) { ThreadItem(it, now, labels[it.workspace_id], onOpenThread) }
        }
        val open = state.workspaces?.items.orEmpty().filter { it.open }
        if (open.isNotEmpty()) {
            header("Workspaces")
            items(open, key = { "workspace:" + it.workspace_id }) { WorkspaceItem(it, onOpenWorkspace) }
        } else if (state.workspaces != null) note("noworkspaces", "No workspaces on this host yet. Add one from Verde on your computer.")
    }
}

@Composable
private fun WorkspaceItem(workspace: Workspace, onOpen: (String) -> Unit) {
    @Suppress("DEPRECATION") val clipboard = LocalClipboardManager.current
    val chats = workspace.threads.count { !subagent(it) && !it.archived }
    val terminals = workspace.panes.count { it.kind == "terminal" && it.terminal_id != null }
    val active = workspace.panes.count { it.attention || it.can_stop || it.status == "working" }
    val flagged = workspace.panes.count { it.attention_kind != null }
    val summary = listOfNotNull(
        "$chats chat${if (chats == 1) "" else "s"}",
        terminals.takeIf { it > 0 }?.let { "$it terminal${if (it == 1) "" else "s"}" },
        active.takeIf { it > 0 }?.let { "$it active" },
        "closed".takeIf { !workspace.open },
    ).joinToString(" · ")
    MenuRow(workspace.label, listOf(summary, workspace.path).filter { it.isNotEmpty() }.joinToString("\n"), true,
        { onOpen(workspace.workspace_id) },
        listOf("Open" to { onOpen(workspace.workspace_id) }, "Copy path" to { clipboard.setText(AnnotatedString(workspace.path)) }),
        leading = { Dot(if (active > 0) MaterialTheme.colorScheme.primary else MaterialTheme.colorScheme.outline, if (active > 0) "Active" else "Idle") },
        badge = flagged.takeIf { it > 0 }?.let { "$it need${if (it == 1) "s" else ""} attention" })
}

@Composable
internal fun WorkspacesScreen(model: BrowseModel, onOpenWorkspace: (String) -> Unit, onHosts: () -> Unit, onPair: () -> Unit) {
    val state by model.state.collectAsState()
    val now = rememberNow(false)
    BrowseFrame("Workspaces", state, now, model::refresh, onHosts) {
        if (!gate(state, onPair)) return@BrowseFrame
        val all = state.workspaces?.items.orEmpty()
        if (all.isEmpty()) { note("empty", "No workspaces on this host yet. Add one from Verde on your computer."); return@BrowseFrame }
        items(all.filter { it.open }, key = { it.workspace_id }) { WorkspaceItem(it, onOpenWorkspace) }
        val closed = all.filterNot { it.open }
        if (closed.isNotEmpty()) {
            header("Closed")
            items(closed, key = { "closed:" + it.workspace_id }) { WorkspaceItem(it, onOpenWorkspace) }
        }
    }
}

@Composable
internal fun WorkspaceScreen(
    model: BrowseModel,
    workspaceId: String,
    onOpenPane: (Pane) -> Unit,
    onOpenThread: (ThreadSummary) -> Unit,
    onHosts: () -> Unit,
    onPair: () -> Unit,
    onNewTerminal: (() -> Unit)? = null,
    onBack: () -> Unit,
) {
    val state by model.state.collectAsState()
    val workspace = state.workspaces?.items?.find { it.workspace_id == workspaceId }
    val now = rememberNow(workspace?.panes.orEmpty().any { it.can_stop && it.started_at_ms != null })
    var showArchived by rememberSaveable(workspaceId) { mutableStateOf(false) }
    BrowseFrame(workspace?.label ?: "Workspace", state, now, model::refresh, onHosts, onBack) {
        if (!gate(state, onPair)) return@BrowseFrame
        if (workspace == null) { note("missing", "This workspace is no longer on the host."); return@BrowseFrame }
        if (workspace.path.isNotEmpty()) note("path", workspace.path)
        header("Panes")
        if (workspace.panes.isEmpty()) note("nopanes", "No open panes.")
        items(workspace.panes, key = { "pane:" + it.id }) { PaneItem(it, now, onOpenPane) }
        if (onNewTerminal != null && canWrite(state.host) && workspace.path.isNotEmpty()) item(key = "new-terminal") {
            TextButton(onClick = onNewTerminal, modifier = Modifier.padding(horizontal = 8.dp)) { Text("New terminal") }
        }
        val threads = workspaceThreads(workspace)
        val (archived, current) = threads.partition { it.archived }
        header("Chats")
        if (current.isEmpty()) note("nochats", "No chats in this workspace yet.")
        items(current, key = { "thread:" + it.thread_id }) { ThreadItem(it, now, null, onOpenThread) }
        if (archived.isNotEmpty()) {
            item(key = "archived-toggle") {
                TextButton(onClick = { showArchived = !showArchived }, modifier = Modifier.padding(horizontal = 8.dp)) {
                    Text(if (showArchived) "Hide archived (${archived.size})" else "Show archived (${archived.size})")
                }
            }
            if (showArchived) items(archived, key = { "archived:" + it.thread_id }) { ThreadItem(it, now, null, onOpenThread) }
        }
    }
}

File: packages/mobile_ios/App/BrowseScreens.swift
import SwiftUI
import UIKit

enum BrowseRoute: Hashable {
    case workspace(String)
    case thread(workspace: String, thread: String)
    case terminal(workspace: String, terminal: String)
}

func route(_ pane: Pane) -> BrowseRoute? {
    guard openable(pane) else { return nil }
    if pane.kind == "terminal", let id = pane.terminal_id { return .terminal(workspace: pane.workspace_id, terminal: id) }
    if let id = pane.thread_id { return .thread(workspace: pane.workspace_id, thread: id) }
    return nil
}

private struct OpenRouteKey: EnvironmentKey {
    static let defaultValue: (BrowseRoute) -> Void = { _ in }
}

extension EnvironmentValues {
    /// Pushes onto the enclosing browse stack (context-menu "Open").
    var openRoute: (BrowseRoute) -> Void {
        get { self[OpenRouteKey.self] }
        set { self[OpenRouteKey.self] = newValue }
    }
}

private func nowMs(_ date: Date) -> Int64 { Int64(date.timeIntervalSince1970 * 1000) }

/// Ticks every second only while a running timer is visible.
private struct Clock<Content: View>: View {
    let ticking: Bool
    @ViewBuilder let content: (Int64) -> Content
    var body: some View {
        TimelineView(.periodic(from: .now, by: ticking ? 1 : 60)) { context in content(nowMs(context.date)) }
    }
}

private func statusColor(_ status: String, attention: Bool) -> Color {
    if attention || status == "waiting_approval" { return .red }
    if ["working", "running", "accepted", "waiting"].contains(status) { return .accentColor }
    return .secondary
}

private struct Dot: View {
    let color: Color
    let label: String
    var body: some View {
        Circle().fill(color).frame(width: 10, height: 10).accessibilityLabel(label)
    }
}

private struct AttentionBadge: View {
    let text: String
    var body: some View {
        Text(text).font(.caption.weight(.semibold)).padding(.horizontal, 6).padding(.vertical, 2)
            .background(Color.red.opacity(0.15), in: Capsule()).foregroundStyle(.red)
    }
}

private struct RowContent: View {
    let title: String
    let detail: String?
    let dot: Dot
    var badge: String?
    var body: some View {
        HStack(spacing: 12) {
            dot
            VStack(alignment: .leading, spacing: 2) {
                Text(title).lineLimit(1)
                if let detail { Text(detail).font(.subheadline).foregroundStyle(.secondary).lineLimit(2) }
            }
            Spacer(minLength: 4)
            if let badge { AttentionBadge(text: badge) }
        }
    }
}

/// Opens through the enclosing NavigationStack; the context menu offers Open and Copy.
private struct MenuRow: View {
    let content: RowContent
    let destination: BrowseRoute?
    let copies: [(String, String)]
    @Environment(\.openRoute) private var openRoute
    var body: some View {
        Group {
            if let destination { NavigationLink(value: destination) { content } } else { content }
        }
        .contextMenu {
            if let destination { Button { openRoute(destination) } label: { Label("Open", systemImage: "arrow.forward") } }
            ForEach(copies.indices, id: \.self) { index in
                // Copied text goes to the pasteboard only; it is never logged.
                Button { UIPasteboard.general.string = copies[index].1 } label: { Label(copies[index].0, systemImage: "doc.on.doc") }
            }
        }
    }
}

private struct PaneItem: View {
    let pane: Pane
    let now: Int64
    var body: some View {
        MenuRow(content: RowContent(title: pane.title, detail: paneLine(pane, now),
                                    dot: Dot(color: statusColor(pane.status, attention: pane.attention), label: paneLabel(pane)),
                                    badge: attentionLabel(pane.attention_kind)),
                destination: route(pane), copies: [("Copy title", pane.title)])
    }
}

private struct ThreadItem: View {
    let thread: ThreadSummary
    let now: Int64
    var workspaceLabel: String?
    var body: some View {
        let parts = [workspaceLabel, thread.status == "idle" ? nil : statusLabel(thread.status),
                     thread.last_activity_at_ms.map { agoLabel($0, now) }].compactMap { $0 }
        MenuRow(content: RowContent(title: thread.title, detail: parts.isEmpty ? nil : parts.joined(separator: " · "),
                                    dot: Dot(color: statusColor(thread.status, attention: thread.status == "waiting_approval"),
                                             label: statusLabel(thread.status))),
                destination: .thread(workspace: thread.workspace_id, thread: thread.thread_id),
                copies: [("Copy title", thread.title)])
    }
}

private struct WorkspaceItem: View {
    let workspace: Workspace
    var body: some View {
        let info = workspaceSummary(workspace)
        MenuRow(content: RowContent(title: workspace.label,
                                    detail: [info.summary, workspace.path].filter { !$0.isEmpty }.joined(separator: "\n"),
                                    dot: Dot(color: info.active > 0 ? .accentColor : .secondary, label: info.active > 0 ? "Active" : "Idle"),
                                    badge: info.badge),
                destination: .workspace(workspace.workspace_id), copies: [("Copy path", workspace.path)])
    }
}

func hostDotColor(_ row: HostRow) -> Color {
    let status = hostStatus(row)
    if status == "Connected" { return .green }
    if row.fatal || row.view?.auth_state == "repair_required" || status.hasPrefix("Unreachable") { return .red }
    return .secondary
}

struct BrowseActions {
    var hosts: () -> Void
    var pair: () -> Void
}

/// Host status, banner and pull-to-refresh shared by the browse screens.
private struct BrowseFrame<Content: View>: View {
    let title: String
    let model: BrowseModel
    let actions: BrowseActions
    var ticking = false
    var large = true
    @ViewBuilder let content: (BrowseState, Int64) -> Content

    var body: some View {
        let state = model.state
        Clock(ticking: ticking) { now in
            List {
                if large, let row = state.row {
                    let status = hostStatus(row)
                    HStack(spacing: 8) {
                        Dot(color: hostDotColor(row), label: status)
                        Text("\(row.saved.label) · \(status)").font(.subheadline)
                    }.listRowSeparator(.hidden)
                }
                if let banner = browseBanner(state, now) {
                    Section {
                        VStack(alignment: .leading, spacing: 8) {
                            Text(banner.text).accessibilityIdentifier("browseBanner")
                            if banner.busy { ProgressView().progressViewStyle(.linear) }
                            switch banner.action {
                            case .some(.retry): Button("Retry") { Task { await model.refresh() } }.disabled(state.refreshing)
                            case .some(.hosts): Button("Open hosts", action: actions.hosts)
                            case nil: EmptyView()
                            }
                        }
                    }.listRowBackground(banner.error ? Color.red.opacity(0.12) : Color.secondary.opacity(0.12))
                }
                if gateOpen(state) {
                    content(state, now)
                } else {
                    Gate(state: state, pair: actions.pair)
                }
            }
            .refreshable { await model.refresh() }
        }
        .navigationTitle(title)
        .navigationBarTitleDisplayMode(large ? .large : .inline)
    }
}

private func gateOpen(_ state: BrowseState) -> Bool {
    state.hostID != nil && !needsPairing(state) && hasContent(state)
}

/// Shared host-level gates shown instead of screen content.
private struct Gate: View {
    let state: BrowseState
    let pair: () -> Void
    var body: some View {
        if state.hostID == nil {
            Text("No host selected.")
            Button("Choose a host", action: pair)
        } else if needsPairing(state) {
            Text("This phone isn't paired with \(state.row?.saved.label ?? "this host") yet.")
            Button("Pair with this host", action: pair).accessibilityIdentifier("pairHost")
        } else if showSpinner(state) {
            HStack { Spacer(); ProgressView("Loading workspaces…"); Spacer() }.padding(.vertical, 24)
        } else {
            Text("No data yet. Pull down to retry when the host is reachable.")
        }
    }
}

struct HomeScreen: View {
    let model: BrowseModel
    let actions: BrowseActions
    var body: some View {
        let panes = model.state.home?.items ?? []
        BrowseFrame(title: "Home", model: model, actions: actions,
                    ticking: panes.contains { $0.can_stop && $0.started_at_ms != nil }) { state, now in
            let attention = panes.filter(\.attention)
            let running = panes.filter { !$0.attention }
            if !attention.isEmpty {
                Section("Needs attention") { ForEach(attention, id: \.id) { PaneItem(pane: $0, now: now) } }
            }
            if !running.isEmpty {
                Section("Running") { ForEach(running, id: \.id) { PaneItem(pane: $0, now: now) } }
            }
            if panes.isEmpty { Section { Text("Nothing is running or waiting on you.") } }
            let workspaces = state.workspaces?.items ?? []
            let labels = Dictionary(workspaces.map { ($0.workspace_id, $0.label) }, uniquingKeysWith: { a, _ in a })
            let recent = recentThreads(state.workspaces)
            if !recent.isEmpty {
                Section("Recent chats") {
                    ForEach(recent, id: \.thread_id) { ThreadItem(thread: $0, now: now, workspaceLabel: labels[$0.workspace_id]) }
                }
            }
            let open = workspaces.filter(\.open)
            if !open.isEmpty {
                Section("Workspaces") { ForEach(open, id: \.workspace_id) { WorkspaceItem(workspace: $0) } }
            } else if state.workspaces != nil {
                Section { Text("No workspaces on this host yet. Add one from Verde on your computer.") }
            }
        }
    }
}

struct WorkspacesScreen: View {
    let model: BrowseModel
    let actions: BrowseActions
    var body: some View {
        BrowseFrame(title: "Workspaces", model: model, actions: actions) { state, _ in
            let all = state.workspaces?.items ?? []
            if all.isEmpty {
                Text("No workspaces on this host yet. Add one from Verde on your computer.")
            } else {
                Section { ForEach(all.filter(\.open), id: \.workspace_id) { WorkspaceItem(workspace: $0) } }
                let closed = all.filter { !$0.open }
                if !closed.isEmpty {
                    Section("Closed") { ForEach(closed, id: \.workspace_id) { WorkspaceItem(workspace: $0) } }
                }
            }
        }
    }
}

struct WorkspaceScreen: View {
    let model: BrowseModel
    let workspaceID: String
    let actions: BrowseActions
    @State private var showArchived = false
    var body: some View {
        let workspace = model.state.workspaces?.items.first { $0.workspace_id == workspaceID }
        BrowseFrame(title: workspace?.label ?? "Workspace", model: model, actions: actions,
                    ticking: workspace?.panes.contains { $0.can_stop && $0.started_at_ms != nil } ?? false,
                    large: false) { _, now in
            if let workspace {
                if !workspace.path.isEmpty { Text(workspace.path).font(.footnote).foregroundStyle(.secondary) }
                Section("Panes") {
                    if workspace.panes.isEmpty { Text("No open panes.") }
                    ForEach(workspace.panes, id: \.id) { PaneItem(pane: $0, now: now) }
                }
                let threads = workspaceThreads(workspace)
                let current = threads.filter { !$0.archived }
                let archived = threads.filter(\.archived)
                Section("Chats") {
                    if current.isEmpty { Text("No chats in this workspace yet.") }
                    ForEach(current, id: \.thread_id) { ThreadItem(thread: $0, now: now) }
                    if !archived.isEmpty {
                        Button(showArchived ? "Hide archived (\(archived.count))" : "Show archived (\(archived.count))") {
                            showArchived.toggle()
                        }
                        if showArchived { ForEach(archived, id: \.thread_id) { ThreadItem(thread: $0, now: now) } }
                    }
                }
            } else {
                Text("This workspace is no longer on the host.")
            }
        }
    }
}

/// D-06 replaces this with the transcript; it deliberately sends no focus/open intents yet.
struct ThreadPlaceholderScreen: View {
    let model: BrowseModel
    let workspaceID: String
    let threadID: String
    var body: some View {
        let state = model.state
        let workspace = state.workspaces?.items.first { $0.workspace_id == workspaceID }
        let thread = workspace?.threads.first { $0.thread_id == threadID }
        let pane = ((state.home?.items ?? []) + (workspace?.panes ?? []))
            .first { $0.workspace_id == workspaceID && $0.thread_id == threadID }
        Clock(ticking: pane?.can_stop == true && pane?.started_at_ms != nil) { now in
            List {
                if let workspace { Text(workspace.label).font(.headline) }
                if let thread {
                    Text([thread.provider, thread.model].compactMap { $0 }.joined(separator: " · "))
                    if let last = thread.last_activity_at_ms { Text("Last activity \(agoLabel(last, now))") }
                }
                Text(pane.map { paneLine($0, now) } ?? statusLabel(thread?.status ?? "idle"))
                if thread == nil && pane == nil { Text("This chat is no longer on the host.") }
                Text("The transcript view is coming in the next update. Open this chat in Verde on your computer for now.")
                    .foregroundStyle(.secondary)
            }
        }
        .navigationTitle(thread?.title ?? pane?.title ?? "Chat")
        .navigationBarTitleDisplayMode(.inline)
    }
}

struct TerminalPlaceholderScreen: View {
    let model: BrowseModel
    let workspaceID: String
    let terminalID: String
    var body: some View {
        let state = model.state
        let workspace = state.workspaces?.items.first { $0.workspace_id == workspaceID }
        let pane = ((state.home?.items ?? []) + (workspace?.panes ?? []))
            .first { $0.workspace_id == workspaceID && $0.terminal_id == terminalID }
        List {
            if let workspace { Text(workspace.label).font(.headline) }
            Text(pane.map { paneLine($0, 0) } ?? "This terminal is no longer on the host.")
            Text("The terminal view is coming in a later update.").foregroundStyle(.secondary)
        }
        .navigationTitle(pane?.title ?? "Terminal")
        .navigationBarTitleDisplayMode(.inline)
    }
}

/// One tab's stack: every browse route resolves against the selected host only.
struct BrowseStack<Root: View>: View {
    let model: BrowseModel
    let actions: BrowseActions
    @Binding var path: [BrowseRoute]
    @ViewBuilder let root: () -> Root
    var body: some View {
        NavigationStack(path: $path) {
            root().navigationDestination(for: BrowseRoute.self) { route in
                switch route {
                case .workspace(let id): WorkspaceScreen(model: model, workspaceID: id, actions: actions)
                case .thread(let workspace, let thread): ThreadPlaceholderScreen(model: model, workspaceID: workspace, threadID: thread)
                case .terminal(let workspace, let terminal): TerminalPlaceholderScreen(model: model, workspaceID: workspace, terminalID: terminal)
                }
            }
        }
        .environment(\.openRoute, { route in path.append(route) })
    }
}
